### ISSUE TITLE
java.sql.SQLRecoverableException: No more data to read from socket

### ISSUE THREAD

**Issue by stevenrigbyS4N** [dnvriend/akka-persistence-jdbc#9]
Hello,

The project I'm currently working on uses akka cluster sharding with this plugin connected to an Oracle database. It works fine but after leaving the project running for some time or when the database is stopped it throws the following error and it is unable to recover:

2014/12/05 09:25:23.887 ERROR[t-dispatcher-14] s.StatementExecutor$$anon$1 81 --- SQL execution failed (Reason: No more data to read from socket):

   MERGE INTO USERSIS.snapshot snapshot USING (SELECT '/user/sharding/RequestIDRouterCoordinator/singleton/coordinator' AS persistence_id, 7 AS seq_nr from DUAL) val ON (snapshot.persistence_id = val.persistence_id and snapshot.sequence_nr = val.seq_nr) WHEN MATCHED THEN UPDATE SET snapshot='qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (1268)' WHEN NOT MATCHED THEN INSERT (PERSISTENCE_ID, SEQUENCE_NR, SNAPSHOT, CREATED) VALUES ('/user/sharding/RequestIDRouterCoordinator/singleton/coordinator', 7, 'qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (1268)', 1417789523876)

2014/12/05 09:25:23.888 ERROR[n-dispatcher-44] s.StatementExecutor$$anon$1 81 --- SQL execution failed (Reason: No more data to read from socket):

   DELETE FROM USERSIS.snapshot WHERE persistence_id = '/user/sharding/RequestIDRouterCoordinator/singleton/coordinator' AND sequence_nr = 7

[ERROR] [12/05/2014 09:25:23.889] [TokenCluster-akka.actor.default-dispatcher-14] [akka://TokenCluster/system/snapshot-store] No more data to read from socket
java.sql.SQLRecoverableException: No more data to read from socket
	at oracle.jdbc.driver.T4CMAREngine.unmarshalUB1(T4CMAREngine.java:1157)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:350)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:227)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:531)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:208)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForRows(T4CPreparedStatement.java:1046)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1336)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3613)
	at oracle.jdbc.driver.OraclePreparedStatement.executeUpdate(OraclePreparedStatement.java:3694)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeUpdate(OraclePreparedStatementWrapper.java:1354)
	at org.apache.commons.dbcp.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:105)
	at org.apache.commons.dbcp.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:105)
	at scalikejdbc.StatementExecutor$$anonfun$executeUpdate$1.apply$mcI$sp(StatementExecutor.scala:337)
	at scalikejdbc.StatementExecutor$$anonfun$executeUpdate$1.apply(StatementExecutor.scala:337)
	at scalikejdbc.StatementExecutor$$anonfun$executeUpdate$1.apply(StatementExecutor.scala:337)
	at scalikejdbc.StatementExecutor$NakedExecutor.apply(StatementExecutor.scala:33)
	at scalikejdbc.StatementExecutor$$anon$1.scalikejdbc$StatementExecutor$LoggingSQLAndTiming$$super$apply(StatementExecutor.scala:317)
	at scalikejdbc.StatementExecutor$LoggingSQLAndTiming$class.apply(StatementExecutor.scala:264)
	at scalikejdbc.StatementExecutor$$anon$1.scalikejdbc$StatementExecutor$LoggingSQLIfFailed$$super$apply(StatementExecutor.scala:317)
	at scalikejdbc.StatementExecutor$LoggingSQLIfFailed$class.apply(StatementExecutor.scala:295)
	at scalikejdbc.StatementExecutor$$anon$1.apply(StatementExecutor.scala:317)
	at scalikejdbc.StatementExecutor.executeUpdate(StatementExecutor.scala:337)
	at scalikejdbc.DBSession$$anonfun$updateWithFilters$1.apply(DBSession.scala:352)
	at scalikejdbc.DBSession$$anonfun$updateWithFilters$1.apply(DBSession.scala:350)
	at scalikejdbc.LoanPattern$class.using(LoanPattern.scala:33)
	at scalikejdbc.ActiveSession.using(DBSession.scala:457)
	at scalikejdbc.DBSession$class.updateWithFilters(DBSession.scala:349)
	at scalikejdbc.ActiveSession.updateWithFilters(DBSession.scala:457)
	at scalikejdbc.DBSession$class.updateWithFilters(DBSession.scala:327)
	at scalikejdbc.ActiveSession.updateWithFilters(DBSession.scala:457)
	at scalikejdbc.SQLUpdate$$anonfun$10.apply(SQL.scala:486)
	at scalikejdbc.SQLUpdate$$anonfun$10.apply(SQL.scala:486)
	at scalikejdbc.DBConnection$class.autoCommit(DBConnection.scala:183)
	at scalikejdbc.DB.autoCommit(DB.scala:75)
	at scalikejdbc.DB$$anonfun$autoCommit$1.apply(DB.scala:218)
	at scalikejdbc.DB$$anonfun$autoCommit$1.apply(DB.scala:217)
	at scalikejdbc.LoanPattern$class.using(LoanPattern.scala:33)
	at scalikejdbc.DB$.using(DB.scala:150)
	at scalikejdbc.DB$.autoCommit(DB.scala:217)
	at scalikejdbc.SQLUpdate.apply(SQL.scala:486)
	at akka.persistence.jdbc.snapshot.GenericStatements$class.deleteSnapshot(Statements.scala:30)
	at akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.deleteSnapshot(SnapshotStores.scala:16)
	at akka.persistence.jdbc.snapshot.JdbcSyncSnapshotStore$class.delete(JdbcSyncSnapshotStore.scala:34)
	at akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.delete(SnapshotStores.scala:16)
	at akka.persistence.snapshot.SnapshotStore$$anonfun$receive$1.applyOrElse(SnapshotStore.scala:44)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.aroundReceive(SnapshotStores.scala:16)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
	at akka.dispatch.Mailbox.run(Mailbox.scala:220)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

It seems that it is unable to detect that the connection has expired or has been closed. 

I'm not sure if this is the correct place to post this.

**Comment by dnvriend** [comment]
Should be fixed when using the validationQuery setting v1.0.9

**Comment by bpg** [comment]
Tested on 1.0.9 version, the same errror:
```
8:52:26.879 [isb-akka.actor.default-dispatcher-25] ERROR s.StatementExecutor$$anon$1 - SQL execution failed (Reason: No more data to read from socket):

   MERGE INTO akka_snapshot snapshot USING (SELECT 'DispatcherActor' AS persistence_id, 4 AS seq_nr from DUAL) val ON (snapshot.persistence_id = val.persistence_id and snapshot.sequence_nr = val.seq_nr) WHEN MATCHED THEN UPDATE SET snapshot='qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (3900)' WHEN NOT MATCHED THEN INSERT (PERSISTENCE_ID, SEQUENCE_NR, SNAPSHOT, CREATED) VALUES ('DispatcherActor', 4, 'qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (3900)', 1421884345964)
```

I can confirm than the new option `validationQuery` works, as a see my query `SELECT 1 FROM DUAL` as `cfg.validationQuery` value in `ScalikeExtensionImpl`:
```
  val connectionPoolSettings = new ConnectionPoolSettings(validationQuery = cfg.validationQuery)
  ConnectionPool.singleton(cfg.url, cfg.username, cfg.password, connectionPoolSettings)
```

In my test scenario application drops and re-creates `akka_snapshot` and `akka_journal` tables at start, then it works for ~40 seconds and gives me that error.
During this 40 second interval it managed to create 4 journal record and 1 snapshot:
```
18:51:48.731 [isb-akka.actor.default-dispatcher-21] DEBUG a.p.j.s.OracleSyncSnapshotStore - loading for persistenceId: DispatcherActor, criteria: SnapshotSelectionCriteria(9223372036854775807,9223372036854775807)
18:51:48.912 [isb-akka.actor.default-dispatcher-21] DEBUG a.p.j.journal.OracleSyncWriteJournal - Async replay for processorId [DispatcherActor], from sequenceNr: [1], to sequenceNr: [9223372036854775807] with max records: [9223372036854775807]
18:51:48.923 [isb-akka.actor.default-dispatcher-21] DEBUG a.p.j.journal.OracleSyncWriteJournal - Async read for highest sequence number for processorId: [DispatcherActor] (hint, seek from  nr: [0])
18:52:25.767 [isb-akka.actor.default-dispatcher-36] DEBUG a.p.j.journal.OracleSyncWriteJournal - writeMessages for 1 presistent messages
18:52:25.885 [isb-akka.actor.default-dispatcher-33] DEBUG a.p.j.journal.OracleSyncWriteJournal - writeMessages for 1 presistent messages
18:52:25.936 [isb-akka.actor.default-dispatcher-33] DEBUG a.p.j.journal.OracleSyncWriteJournal - writeMessages for 1 presistent messages
18:52:25.948 [isb-akka.actor.default-dispatcher-33] DEBUG a.p.j.journal.OracleSyncWriteJournal - writeMessages for 1 presistent messages
18:52:25.966 [isb-akka.actor.default-dispatcher-36] DEBUG a.p.j.s.OracleSyncSnapshotStore - Saving metadata: SnapshotMetadata(DispatcherActor,4,1421884345964), snapshot: DispatcherState@226117a1
18:52:26.881 [isb-akka.actor.default-dispatcher-25] DEBUG a.p.j.s.OracleSyncSnapshotStore - Deleting: SnapshotMetadata(DispatcherActor,4,0)
```
and then failed on the last delete.

**Comment by bpg** [comment]
I did some research on this topic, and I found that in configuration that I currently have (the default, I guess) scalikejdbc 2.2.2 uses commons-dbcp v.1.4, and the root cause of the error is exception `java.sql.SQLException: OALL8 is in an inconsistent state` the happens there in `PoolableConnectionFactory.validateObject`

Googling for this exception gave me a bunch of articles suggested to update DB driver. In my deployment it was Oracle 11.2 XE server and  ojdbc14 v10.2.0.3.0 driver. I upgraded driver to 11.2.0.4, but but now I have a different exception: `No more data to read from socket` 
:confused:

**Comment by dnvriend** [comment]
Clone the latest version 1.0.9-SNAPSHOT, built it locally using sbt clean +publish-m2 and try it on your project. I added 2 tests, restarted the oracle docker and still the connections will not drop. The errors I get is when I restart oracle, the database does not accept connections (when the test wants to run and the db is not ready), but this is obviously correct. I changed the connection pool to commons-dbcp, the default and changed the minimum connections from 0 to 1, maybe this will help. Also try the driver from the lib directory in the project, the ojdbc6.jar.

**Comment by bpg** [comment]
Hi Dennis,

Thanks for the prompt reply.
I tried all you suggested:
- 1.0.9-SNAPSHOT
- ojdbc6 from the project's lib

and
- the Oracle XE instance from docker image you're using in tests
- a separate user / schema dedicated for the journal and snapshot tables
- monitored number of used connections, it was always < max allowed connections
- recipe from https://community.oracle.com/message/3701989
 
result is the same:   
`20:50:36.299 [isb-akka.actor.default-dispatcher-3] ERROR s.StatementExecutor$$anon$1 - SQL execution failed (Reason: IO Error: Connection reset):`

Below is akka-persistence-jdbc debug log:
```
20:49:32.856 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-26] DEBUG scalikejdbc.ConnectionPool$ - Registered connection pool : ConnectionPool(url:jdbc:oracle:thin:@boot2docker:1521:, user:PSYS_AKKA) using factory : commons-dbcp
20:49:32.858 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-26] DEBUG scalikejdbc.ConnectionPool$ - Registered singleton connection pool : ConnectionPool(url:jdbc:oracle:thin:@boot2docker:1521:, user:PSYS_AKKA)
20:49:33.019 [isb-akka.actor.default-dispatcher-21] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   SELECT * FROM PSYS_AKKA.akka_snapshot WHERE persistence_id = 'DispatcherActor' AND sequence_nr <= 9223372036854775807 ORDER BY sequence_nr DESC; (12 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.snapshot.GenericStatements$class.selectSnapshotsFor(Statements.scala:52)
    akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.selectSnapshotsFor(SnapshotStores.scala:16)
    akka.persistence.jdbc.snapshot.JdbcSyncSnapshotStore$$anonfun$loadAsync$1.apply(JdbcSyncSnapshotStore.scala:21)
    akka.persistence.jdbc.snapshot.JdbcSyncSnapshotStore$$anonfun$loadAsync$1.apply(JdbcSyncSnapshotStore.scala:21)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
    akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
    scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
    ...

20:49:33.063 [isb-akka.actor.default-dispatcher-21] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   SELECT message FROM PSYS_AKKA.akka_journal WHERE persistence_id = 'DispatcherActor' AND (sequence_number >= 1 AND sequence_number <= 9223372036854775807) AND ROWNUM <= 9223372036854775807 ORDER BY sequence_number; (9 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.OracleStatements$class.selectMessagesFor(Statements.scala:102)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.selectMessagesFor(Journals.scala:16)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$asyncReplayMessages$1.apply$mcV$sp(JdbcSyncWriteJournal.scala:78)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$asyncReplayMessages$1.apply(JdbcSyncWriteJournal.scala:78)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$asyncReplayMessages$1.apply(JdbcSyncWriteJournal.scala:78)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
    akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
    scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
    ...

20:49:33.083 [isb-akka.actor.default-dispatcher-21] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   SELECT MAX(sequence_number) FROM PSYS_AKKA.akka_journal WHERE persistence_id = 'DispatcherActor'; (3 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.GenericStatements$$anonfun$selectMaxSequenceNr$1.apply$mcJ$sp(Statements.scala:67)
    akka.persistence.jdbc.journal.GenericStatements$$anonfun$selectMaxSequenceNr$1.apply(Statements.scala:69)
    akka.persistence.jdbc.journal.GenericStatements$$anonfun$selectMaxSequenceNr$1.apply(Statements.scala:69)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
    akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
    scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
    ...

20:50:35.812 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-32] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   INSERT INTO PSYS_AKKA.akka_journal (persistence_id, sequence_number, marker, message, created) VALUES ('DispatcherActor',1,'A','CucSCAES4hKs7QAFc3IATWNvbS5wZXJzcGVjc3lzLmlzYi5pbXBsLmFra2Euam9iZXhlY3V0b3IuZGlzcGF0Y2hlci5ldmVudC5O... (3344)', current_timestamp); (12 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.GenericStatements$class.insertMessage(Statements.scala:43)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.insertMessage(Journals.scala:16)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:24)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:22)
    scala.collection.Iterator$class.foreach(Iterator.scala:727)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$class.writeMessages(JdbcSyncWriteJournal.scala:22)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.writeMessages(Journals.scala:16)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply$mcV$sp(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    scala.util.Try$.apply(Try.scala:161)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1.applyOrElse(SyncWriteJournal.scala:27)
    ...

20:50:36.022 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-32] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   INSERT INTO PSYS_AKKA.akka_journal (persistence_id, sequence_number, marker, message, created) VALUES ('DispatcherActor',2,'A','CsMCCAESvgKs7QAFc3IARWNvbS5wZXJzcGVjc3lzLmlzYi5pbXBsLmFra2Euam9iZXhlY3V0b3IuZGlzcGF0Y2hlci5ldmVudC5H... (716)', current_timestamp); (5 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.GenericStatements$class.insertMessage(Statements.scala:43)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.insertMessage(Journals.scala:16)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:24)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:22)
    scala.collection.Iterator$class.foreach(Iterator.scala:727)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$class.writeMessages(JdbcSyncWriteJournal.scala:22)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.writeMessages(Journals.scala:16)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply$mcV$sp(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    scala.util.Try$.apply(Try.scala:161)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1.applyOrElse(SyncWriteJournal.scala:27)
    ...

20:50:36.056 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-32] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   INSERT INTO PSYS_AKKA.akka_journal (persistence_id, sequence_number, marker, message, created) VALUES ('DispatcherActor',3,'A','CogECAESgwSs7QAFc3IAS2NvbS5wZXJzcGVjc3lzLmlzYi5pbXBsLmFra2Euam9iZXhlY3V0b3IuZGlzcGF0Y2hlci5ldmVudC5K... (980)', current_timestamp); (3 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.GenericStatements$class.insertMessage(Statements.scala:43)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.insertMessage(Journals.scala:16)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:24)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:22)
    scala.collection.Iterator$class.foreach(Iterator.scala:727)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$class.writeMessages(JdbcSyncWriteJournal.scala:22)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.writeMessages(Journals.scala:16)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply$mcV$sp(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    scala.util.Try$.apply(Try.scala:161)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1.applyOrElse(SyncWriteJournal.scala:27)
    ...

20:50:36.072 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-32] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   INSERT INTO PSYS_AKKA.akka_journal (persistence_id, sequence_number, marker, message, created) VALUES ('DispatcherActor',4,'A','CucSCAES4hKs7QAFc3IATWNvbS5wZXJzcGVjc3lzLmlzYi5pbXBsLmFra2Euam9iZXhlY3V0b3IuZGlzcGF0Y2hlci5ldmVudC5O... (3404)', current_timestamp); (6 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.journal.GenericStatements$class.insertMessage(Statements.scala:43)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.insertMessage(Journals.scala:16)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:24)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$$anonfun$writeMessages$1.apply(JdbcSyncWriteJournal.scala:22)
    scala.collection.Iterator$class.foreach(Iterator.scala:727)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    akka.persistence.jdbc.journal.JdbcSyncWriteJournal$class.writeMessages(JdbcSyncWriteJournal.scala:22)
    akka.persistence.jdbc.journal.OracleSyncWriteJournal.writeMessages(Journals.scala:16)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply$mcV$sp(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1$$anonfun$1.apply(SyncWriteJournal.scala:27)
    scala.util.Try$.apply(Try.scala:161)
    akka.persistence.journal.SyncWriteJournal$$anonfun$receive$1.applyOrElse(SyncWriteJournal.scala:27)
    ...

20:50:36.299 [isb-akka.actor.default-dispatcher-3] ERROR s.StatementExecutor$$anon$1 - SQL execution failed (Reason: IO Error: Connection reset):

   MERGE INTO PSYS_AKKA.akka_snapshot snapshot USING (SELECT 'DispatcherActor' AS persistence_id, 4 AS seq_nr from DUAL) val ON (snapshot.persistence_id = val.persistence_id and snapshot.sequence_nr = val.seq_nr) WHEN MATCHED THEN UPDATE SET snapshot='qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (3900)' WHEN NOT MATCHED THEN INSERT (PERSISTENCE_ID, SEQUENCE_NR, SNAPSHOT, CREATED) VALUES ('DispatcherActor', 4, 'qAAAAKztAAVzcgAtYWtrYS5wZXJzaXN0ZW5jZS5zZXJpYWxpemF0aW9uLlNuYXBzaG90SGVhZGVyAAAAAAAAAAECAAJJAAxzZXJp... (3900)', 1421977836080)

20:50:39.635 [isb-akka.persistence.dispatchers.default-plugin-dispatcher-33] DEBUG s.StatementExecutor$$anon$1 - SQL execution completed

  [SQL Execution]
   DELETE FROM PSYS_AKKA.akka_snapshot WHERE persistence_id = 'DispatcherActor' AND sequence_nr = 4; (6 ms)

  [Stack Trace]
    ...
    akka.persistence.jdbc.snapshot.GenericStatements$class.deleteSnapshot(Statements.scala:30)
    akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.deleteSnapshot(SnapshotStores.scala:16)
    akka.persistence.jdbc.snapshot.JdbcSyncSnapshotStore$class.delete(JdbcSyncSnapshotStore.scala:34)
    akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.delete(SnapshotStores.scala:16)
    akka.persistence.snapshot.SnapshotStore$$anonfun$receive$1.applyOrElse(SnapshotStore.scala:44)
    akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    akka.persistence.jdbc.snapshot.OracleSyncSnapshotStore.aroundReceive(SnapshotStores.scala:16)
    akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    akka.actor.ActorCell.invoke(ActorCell.scala:487)
    akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
    akka.dispatch.Mailbox.run(Mailbox.scala:221)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    java.lang.Thread.run(Thread.java:745)
    ...

```

I've never been an Oracle guru, but it seems that `MERGE... ` statement causes the problem. Execution fails always on MERGE. Apparently, the next `DELETE` statement passes through. 
My application executes some other SQLs to this schema in parallel without any hiccups.
What do you think, could we use some other combination of SQL statements instead of MERGE to achieve the same functionality?

**Comment by dnvriend** [comment]
Did adding the following setting when launching the JVM help, as suggested by (https://community.oracle.com/message/3701989) ?

```
java -jar your.jar -Djava.security.egd=file:/dev/../dev/urandom your.main.Class
```

**Comment by bpg** [comment]
No, setting property `java.security.egd` did not help. I'm running this on Mac (as well as you, I assume)

**Comment by bpg** [comment]
As I suspected, the problem is in `MERGE` statement when the data object stored in snapshot is quite big. I made some changes in unit tests to reproduce it, could you please take a look at  bpg/akka-persistence-jdbc@bfa643bf334957c4dc01895839ba5eabb8f0be2e ?
Thanks.

**Comment by bpg** [comment]
Reverted usage of `MERGE`. Now akka-persistence-jdbc works seamlessly on my project!

**Comment by dnvriend** [comment]
Hi

Good find the merge statement! ☺ Makes you wonder why that is. I will look at the code later this week and 'merge' it...

regards

**Comment by bpg** [comment]
Let me make a pull request to your repo.
I replaced `MERGE` with t-sql stored procedure call. Just need to update unit tests to make sure everything is OK, will be ready by tomorrow.

**Comment by dnvriend** [comment]
also makes you wonder about performance.. When the insert fails its basically an extra call.. before the update (recover which is fine) thats why the merge Seemed such a nice solution

**Comment by bpg** [comment]
SP addresses the performance concern as well

**Comment by dnvriend** [comment]
thats true. darn its 04.00 o clock here gonna get some more sleep gotta go to work in 4 hours lol. Looking forward to your pr. Good luck!

**Comment by dnvriend** [comment]
Fixed with release 1.1.0
